diag_manager/fms_diag_file_object.F90

```diff
@@ -120,48 +119,62 @@ module fms_diag_file_object_mod
 #endif
 
 end type fmsDiagFile_type
+type, extends (fmsDiagFile_type) :: subRegionalFile_type
+  integer, dimension(:), allocatable :: sub_axis_ids !< Array of axis ids in the file
+end type subRegionalFile_type
 
-type(fmsDiagFile_type), dimension (:), allocatable, target :: FMS_diag_files !< The array of diag files
+!> \brief A container for fmsDiagFile_type.  This is used to create the array of files
+type fmsDiagFileContainer_type
+  class (fmsDiagFile_type),allocatable :: FMS_diag_file !< The individual file object 
+end type fmsDiagFileContainer_type
 
+!type(fmsDiagFile_type), dimension (:), allocatable, target :: FMS_diag_file !< The array of diag files
+!class(fmsDiagFileContainer_type),dimension (:), allocatable, target :: FMS_diag_file
 contains
 
 !< @brief Allocates the number of files and sets an ID based for each file
 !! @return true if there are files allocated in the YAML object
-logical function fms_diag_files_object_init ()
+logical function fms_diag_files_object_init (files_array)
 #ifdef use_yaml
+  class(fmsDiagFileContainer_type), allocatable, target, intent(inout) :: files_array (:) !< array of diag files
+  class(fmsDiagFile_type), pointer :: obj => null() !< Pointer for each member of the array
   integer :: nFiles !< Number of files in the diag yaml
   integer :: i !< Looping iterator
-  type(fmsDiagFile_type), pointer :: obj !< FMS_diag_files(i) (for less typing)
   if (diag_yaml%has_diag_files()) then
    nFiles = diag_yaml%size_diag_files()
-   allocate (FMS_diag_files(nFiles))
+   allocate (files_array(nFiles))
    set_ids_loop: do i= 1,nFiles
-     obj => FMS_diag_files(i)
+     !> If the file has a sub_regional, define it as one and allocate the sub_axis_ids array.
+     !! This will be set in a add_axes
+     files_array(i)%FMS_diag_file%diag_yaml_file => diag_yaml%diag_files(i)
+     if (files_array(i)%FMS_diag_file%has_file_sub_region()) then
```

Comment:
`files_array(i)%FMS_diag_file` has not been allocated at this point. 

I think you remove L149 and just do
```F90
if (diag_yaml%diag_files(i)%has_file_sub_region()) then
```

`files_array(i)%FMS_diag_file%diag_yaml_file`  is then set in L164 bellow